upup/pkg/fi/cloudup/awstasks/load_balancer_cleanup.go

```diff
@@ -0,0 +1,461 @@
+
+	if err != nil {
+		return fmt.Errorf("Error Detaching LoadBalancer TargetGroup from Autoscaling group : %v", err)
+	}
+
+	return nil
+}
+
+func (d *detachNetworkLoadBalancer) String() string {
+	return d.TaskName() + "-" + d.Item()
+}
+
+func (e *LoadBalancerCleanup) FindELBDeletions(c *fi.Context) ([]fi.Deletion, error) {
+	var removals []fi.Deletion
+
+	cloud := c.Cloud.(awsup.AWSCloud)
+
+	lb, err := FindLoadBalancerByNameTag(cloud, fi.StringValue(e.LBName))
+
+	if err != nil {
+		return nil, err
+	}
+
+	if lb != nil {
+
+		request := &elb.DeleteLoadBalancerInput{
+			LoadBalancerName: lb.LoadBalancerName,
+		}
+
+		removals = append(removals, &deleteLoadBalancer{request: request})
+		klog.V(2).Infof("will delete load balancer: %v", lb.LoadBalancerName)
+	}
+
+	for _, autoScalingGroupName := range e.AgNames {
+
+		request := &autoscaling.DescribeLoadBalancersInput{
+			AutoScalingGroupName: autoScalingGroupName,
+		}
+		response, err := cloud.Autoscaling().DescribeLoadBalancers(request)
+
+		if err != nil || response == nil { //TODO: Mock Autoscaling is returning nil
+			return nil, nil
+		}
+
+		for _, LoadBalancerState := range response.LoadBalancers { //detach all elbs from autoscaling group
+
+			loadBalancerName := LoadBalancerState.LoadBalancerName
+
+			request := &autoscaling.DetachLoadBalancersInput{
+				AutoScalingGroupName: autoScalingGroupName,
+				LoadBalancerNames: []*string{
+					loadBalancerName,
+				},
+			}
+
+			removals = append(removals, &detachLoadBalancer{request: request})
+			klog.V(2).Infof("will detach load balancer: %v from autoscalinggroup %v", loadBalancerName, autoScalingGroupName)
+		}
+	}
+
+	return removals, nil
+}
+
+func (e *LoadBalancerCleanup) FindNLBDeletions(c *fi.Context) ([]fi.Deletion, error) {
+	var removals []fi.Deletion
+
+	cloud := c.Cloud.(awsup.AWSCloud)
+
+	lb, err := FindNetworkLoadBalancerByNameTag(cloud, fi.StringValue(e.LBName))
+
+	if err != nil {
+		return nil, err
+	}
+
+	if lb != nil {
+
+		request := &elbv2.DeleteLoadBalancerInput{
+			LoadBalancerArn: lb.LoadBalancerArn,
+		}
+
+		removals = append(removals, &deleteNetworkLoadBalancer{request: request})
+		klog.V(2).Infof("will delete network load balancer: %v", lb.LoadBalancerName)
+
+		//TODO: Could be useful to check the autoscaling group for associated target groups and delete them as well.
+		{
+			klog.V(2).Infof("Describing Target Groups for loadBalancerArn : %v\n", lb.LoadBalancerArn)
+			request := &elbv2.DescribeTargetGroupsInput{
+				LoadBalancerArn: lb.LoadBalancerArn,
+			}
+			response, err := cloud.ELBV2().DescribeTargetGroups(request)
+			if err != nil {
+				return nil, fmt.Errorf("error querying for NLB Target groups :%v", err)
+			}
+
+			if len(response.TargetGroups) == 0 {
+				return nil, fmt.Errorf("Found no Target Groups for NLB don't think this is a normal condition :  %v", lb.LoadBalancerArn)
+			}
+
+			if len(response.TargetGroups) != 1 {
```

Comment:
Theres a good chance that we'll use multiple target groups in the fix for https://github.com/kubernetes/kops/issues/9756 so we should handle deleting all of them.

Reply:
if we'll use multiple target groups, network_loadbalancer.go L393 will need adjustment.